megamek/src/megamek/client/ui/preferences/JTabbedPanePreference.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2021 - The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MegaMek.
+ *
+ * MegaMek is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * MegaMek is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.client.ui.preferences;
+
+import javax.swing.*;
+import javax.swing.event.ChangeEvent;
+import javax.swing.event.ChangeListener;
+import java.lang.ref.WeakReference;
+
+public class JTabbedPanePreference extends PreferenceElement implements ChangeListener {
+    //region Variable Declarations
+    private final WeakReference<JTabbedPane> weakReference;

```

Comment:
(In general, across these Preference classes), what exactly is this pointing to? An individual TabbedPane/TextBox/etc?

Reply:
Exactly.

I used the previously created basis in MekHQ for these, but the implementation is interesting and described here: https://docs.oracle.com/javase/8/docs/api/java/lang/ref/WeakReference.html

Reply:
I designed this a long time ago :) The preferences system was a way to save UI state between MekHQ sessions, so users would not have to setup the same selections or comboboxes every time they started MekHQ.

The system it's formed by two parts:
- Preferences are saved/read from a .json file. While the app is running, they exist in memory as PreferenceNodes and JXYZPreferences, forming a tree.
- JXYZPreference: these classes wrap a Swing UI element, and have a 1 to 1 relationship with them. Each class implements the relevant listeners so it can update itself when the UI element changes, and it also knows how to update the UI element based on the saved values of the .json file.

Think of each PreferenceNode as a pointer to the dialog, and then the different JXYZPreference as the state of each part of that dialog: the state of comboboxes, dialog sizes, text boxes...

But because JXYZPreference contains a UI element, if we used a regular reference, the UI wouldn't be garbage collected because the GC would find that the UI element has something referencing it. WeakReference is a class designed for this, it means that you can access an element, but you won't stop it from being garbage collected.